services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java

```diff
@@ -0,0 +1,159 @@
+        recoveryAuthnCodeUserInput = formParamsMap.getFirst(RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW);
+
+        if (ObjectUtil.isBlank(recoveryAuthnCodeUserInput)) {
+            authnFlowContext.forceChallenge(createLoginForm(authnFlowContext, true,
+                    RecoveryAuthnCodesUtils.RECOVERY_AUTHN_CODES_INPUT_DEFAULT_ERROR_MESSAGE,
+                    RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW));
+        } else {
+            targetRealm = authnFlowContext.getRealm();
+            authenticatedUser = authnFlowContext.getUser();
+            if (!isDisabledByBruteForce(authnFlowContext, authenticatedUser)) {
+                isValid = this.userCredentialManager.isValid(targetRealm, authenticatedUser,
+                        UserCredentialModel.buildFromBackupAuthnCode(recoveryAuthnCodeUserInput.replace("-", "")));
+                if (!isValid) {
+                    responseChallenge = createLoginForm(authnFlowContext, true,
+                            RecoveryAuthnCodesUtils.RECOVERY_AUTHN_CODES_INPUT_DEFAULT_ERROR_MESSAGE,
+                            RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW);
+                    authnFlowContext.failureChallenge(AuthenticationFlowError.INVALID_CREDENTIALS, responseChallenge);
+                } else {
+                    result = true;
+                    optUserCredentialFound = this.userCredentialManager.getStoredCredentialsByTypeStream(targetRealm,
+                            authenticatedUser, RecoveryAuthnCodesCredentialModel.TYPE).findFirst();
+                    if (optUserCredentialFound.isPresent()) {
+                        recoveryCodeCredentialModel = RecoveryAuthnCodesCredentialModel
+                                .createFromCredentialModel(optUserCredentialFound.get());
+                        if (recoveryCodeCredentialModel.allCodesUsed()) {
+                            this.userCredentialManager.removeStoredCredential(targetRealm, authenticatedUser,
+                                    recoveryCodeCredentialModel.getId());
+                        }
+                    }
+                    if (recoveryCodeCredentialModel == null || recoveryCodeCredentialModel.allCodesUsed()) {
+                        authenticatedUser.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES);
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    protected boolean isDisabledByBruteForce(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser) {
+        boolean bolResult = false;
+        String bruteForceError;
+        Response challengeResponse;
+        bruteForceError = AuthenticatorUtils.getDisabledByBruteForceEventError(authnFlowContext.getProtector(),
+                authnFlowContext.getSession(), authnFlowContext.getRealm(), authenticatedUser);
+        if (bruteForceError != null) {
```

Comment:
How about using early exit here as well?

Reply:
You could also use the occasion to introduce a new overload like `AuthenticatorUtils.getDisabledByBruteForceEventError(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser)` to . This could help to streamline the API usage of a lot of other authenticator implementations.

Reply:
OK.